src/main/scala/org/constellation/primitives/TrieBasedTipService.scala

```diff
@@ -0,0 +1,143 @@
+package org.constellation.primitives
+import org.constellation.DAO
+import org.constellation.consensus.TipData
+import org.constellation.primitives.Schema.{Id, SignedObservationEdge}
+import org.constellation.util.Metrics
+
+import scala.collection.concurrent.TrieMap
+import scala.util.Random
+
+trait ConcurrentTipService {
+
+  def getMinTipHeight()(implicit dao: DAO): Long
+  def toMap: Map[String, TipData]
+  def size: Int
+  def set(tips: Map[String, TipData])
+  def update(checkpointBlock: CheckpointBlock)(implicit dao: DAO)
+  // considerd as private only
+  def put(k: String, v: TipData)(implicit metrics: Metrics): Option[TipData]
+  def pull(
+    readyFacilitators: Map[Id, PeerData]
+  )(implicit metrics: Metrics): Option[(Seq[SignedObservationEdge], Map[Id, PeerData])]
+  def get(k: String): Option[TipData]
+  def remove(k: String)(implicit metrics: Metrics): Unit
+
+}
+
+class TrieBasedTipService(sizeLimit: Int,
+                          maxWidth: Int,
+                          numFacilitatorPeers: Int,
+                          minPeerTimeAddedSeconds: Int)
+    extends ConcurrentTipService {
+
+  private val tips: TrieMap[String, TipData] = TrieMap.empty
+
+  override def set(newTips: Map[String, TipData]): Unit = {
+    tips ++= newTips
```

Comment:
Should this update metrics?

Reply:
It didn't previously, if it should please comment which metric key should be incremented.